cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/LocalGitServer.java

```diff
@@ -0,0 +1,116 @@
+ * the License.
+ */
+
+package io.cdap.cdap.sourcecontrol;
+
+import org.eclipse.jetty.security.ConstraintMapping;
+import org.eclipse.jetty.security.ConstraintSecurityHandler;
+import org.eclipse.jetty.security.HashLoginService;
+import org.eclipse.jetty.security.UserStore;
+import org.eclipse.jetty.security.authentication.BasicAuthenticator;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.ServerConnector;
+import org.eclipse.jetty.servlet.ServletHandler;
+import org.eclipse.jetty.servlet.ServletHolder;
+import org.eclipse.jetty.util.security.Constraint;
+import org.eclipse.jetty.util.security.Password;
+import org.eclipse.jgit.http.server.GitServlet;
+import org.eclipse.jgit.lib.Repository;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nullable;
+import javax.servlet.http.HttpServletRequest;
+
+/**
+ * A local server with authentication that hosts a Git repository for tests.
+ */
+public class LocalGitServer {
+  private final Server server;
+  private final GitServlet gitServlet;
+  private final Repository repository;
+  private final String username;
+  private final String password;
+  private ServerConnector serverConnector;
+  private final  int port;
+
+  /**
+   * Creates a {@link LocalGitServer}
+   * @param username username for authentication.
+   * @param password password for authentication.
+   * @param port port to start the server on. Use 0 for a random port.
+   * @param repository the git repository to serve through the server.
+   */
+  public LocalGitServer(String username, String password, int port, Repository repository) {
+    this.server = new Server();
+    this.gitServlet = new GitServlet();
+    this.repository = repository;
+    this.username = username;
+    this.password = password;
+    this.port = port;
+  }
+
+  @Nullable
+  public String getServerURL() {
+    return serverConnector == null ? null : String.format("http://localhost:%s/", serverConnector.getLocalPort());
+  }
+
+  public void start() throws Exception {
+    // Always return the same repository.
+    gitServlet.setRepositoryResolver((HttpServletRequest req, String name) -> repository);
+    // Git requests try to authenticate as admin by default, so set that as the required
+    // security role for all paths.
+    Set<String> constraintRoles = Collections.singleton("admin");
+
+    String[] roles = {"admin"};
+    Constraint constraint = new Constraint();
+    constraint.setRoles(roles);
+    constraint.setAuthenticate(true);
+    constraint.setName("auth");
+
+    ConstraintMapping constraintMapping = new ConstraintMapping();
+    constraintMapping.setPathSpec("/*");
+    constraintMapping.setConstraint(constraint);
+
+    List<ConstraintMapping> constraintMappings = Collections.singletonList(constraintMapping);
+
+    ConstraintSecurityHandler securityHandler = new ConstraintSecurityHandler();
```

Comment:
some comments about what a specific code block is trying to do will be helpful later

Reply:
Added a some comments.